src/utils.jl

```diff
@@ -13,10 +13,7 @@ nansof(::Type{TM}, n...) where TM = (TM(n...) .= NaN)
 
 @inline function matdot(v::AbstractVector, A::AbstractMatrix, j::Int)
     T = promote_type(eltype(v), eltype(A))
-    r = zero(T)
-    for i in 1:length(v)
-        r += v[i] * A[j, i]
-    end
+    r = mapreduce((va)->(va[1]*va[2]), +, zero(T), zip(v, @view(A[j,:])))
```

Comment:
Will look är this closer later but to me, lines like this is almost inscrutinable. Does it have an advantage over a loop?

Reply:
I tried using `map` and `mapreduce` wherever possible because these are supported functions in [KissThreading.jl](https://github.com/bkamins/KissThreading.jl), [GPUArrays.jl](https://github.com/JuliaGPU/GPUArrays.jl) and [DistributedArrays.jl](https://github.com/JuliaParallel/DistributedArrays.jl/blob/master/src/mapreduce.jl). So by sticking to them, we get support for all these nice arrays. MMA's strongest advantage IMO is its ability to parallelize very easily. I will make some bug fix and add a fallback in KissThreading.jl to avoid the overhead when only 1 thread is available, then I will use it as a dependency here in another PR.